### ISSUE TITLE
Clang-v12 vs GCC v9 generates different stream [portability-issue]

### ISSUE THREAD

**Issue by ph4r05** [imneme/pcg-cpp#67]
According to the GCC standard (§8.3.6): "The order of evaluation of function arguments is unspeciﬁed."
This causes a problem when PCG is initialized from another generator. GNU GCC evaluates from right (v9.3) while Apple Clang evaluates from left (v12).

The PCG is thus non-portable as with the same seed it produced different results on various platforms.
Problematic code:

https://github.com/imneme/pcg-cpp/blob/5b5cac8d61339e810c5dbb4692d868a1d7ca1b2d/include/pcg_random.hpp#L514-L521

Line 520, 521. As we take two draws from the same generator, the order of evaluation matters. Clang thus generates a different stream than gcc. 

It seems desirable to have a portable random number generation. (At least my use-case - scientific computation, experiment reproducibility requires it).

More on this:
- https://en.cppreference.com/w/c/language/eval_order
- https://stackoverflow.com/questions/15440833/g-vs-intel-clang-argument-passing-order

**Comment by imneme** [comment]
Good catch! Sorry about that.

**Comment by ph4r05** [comment]
Thanks @imneme ! I figured it randomly when I was debugging differences of our experiment on my local OSX vs. computing grid (debian10).

I've created demo PR #68 that I am using to fix the problem.

**Comment by imneme** [comment]
I'm not sure I'd fix it the way you did, but I'm still mulling a clean way to do so.

**Comment by imneme** [comment]
I think it might be cleanest/simplest to replace
```
    template<typename SeedSeq>
    engine(SeedSeq&& seedSeq, typename std::enable_if<
                   stream_mixin::can_specify_stream
               && !std::is_convertible<SeedSeq, itype>::value
               && !std::is_convertible<SeedSeq, engine>::value,
        can_specify_stream_tag>::type = {})
        : engine(generate_one<itype,1,2>(seedSeq),
                 generate_one<itype,0,2>(seedSeq))
    {
        // Nothing else to do.
    }
```
with
```
    template<typename SeedSeq>
    engine(SeedSeq&& seedSeq, typename std::enable_if<
                   stream_mixin::can_specify_stream
               && !std::is_convertible<SeedSeq, itype>::value
               && !std::is_convertible<SeedSeq, engine>::value,
        can_specify_stream_tag>::type = {})
    {
        itype seedval = generate_one<itype,1,2>(seedSeq);
        itype stream = generate_one<itype,0,2>(seedSeq));
        seed(seedval, stream);
    }
```
Thoughts?

**Comment by ph4r05** [comment]
@imneme aah this also seems to work! I didn't notice the `seed` method. 

I was also aiming to preserve one way of the data generation so it is not breaking for both GCC and Clang. 

It seems that 

```cpp
generate_to<2>(std::forward<SeedSeq>(seedSeq), seeddata);
```

generates different result as

```cpp
auto s0 = generate_one<itype,0,2>(seedSeq);
auto s1 = generate_one<itype,1,2>(seedSeq);
```

is that correct?

**Comment by ph4r05** [comment]
The following code seems to be equivalent to the old code on GNU GCC

```cpp
itype seeddata[4];
generate_to<4>(std::forward<SeedSeq>(seedSeq), seeddata);
seed(seeddata[3], seeddata[0]);
```

**Comment by imneme** [comment]
BTW, the original code only has undefined behavior for a non-compliant seedseq that returns different data every time you call `generate`.  With a standards compliant `generate` member function, it should all be correct because `generate` will return the same data regardless of call order, as shown in this sample code:

```c++
#include <random>
#include <iostream>
#include <cstdint>

int main()
{
    std::seed_seq seq { 42, 54 };
    constexpr size_t MAX = 6;
    uint32_t seeds[MAX];
    for (int j = 0; j < 10; ++j) {
	seq.generate(seeds, seeds+MAX);
	for (int i = 0; i < MAX; ++i)
	    std::cout << (i > 0 ? ", " : "") << seeds[i];
	std::cout << "\n";
    }
    return 0;
}
```

**Comment by imneme** [comment]
> The following code seems to be equivalent to the old code on GNU GCC
> 
> ```c++
> itype seeddata[4];
> generate_to<4>(std::forward<SeedSeq>(seedSeq), seeddata);
> seed(seeddata[3], seeddata[0]);
> ```

This would only be the case for a non-compliant `generate` method.  A seed sequence that follows the rules, no.

**Comment by ph4r05** [comment]
Ah interesting, the code I inherited uses the mentioned constructor to initialize the generator from another generator (hierarchy of generators). Now I see where is the problem.

**Comment by ph4r05** [comment]
So then it seems like a problem of our seed seq generator. 
Thanks for your time!

Should I close the issue and PR?

**Comment by imneme** [comment]
> Should I close the issue and PR?

It is an interesting point that with a non-compliant SeedSeq, the PCG's behavior is technically undefined.  But I think its fair to say that when you call the PCG constructor with a seed sequence that “works” but doesn't behave reproducibly, PCG works (in practice) but doesn't behave reproducibly.

Maybe adjust the PR and its title.  I think the revised code is probably more efficient (it's wasteful to call `generate` twice), so I'll probably go with it anyway.

**Comment by ph4r05** [comment]
thanks @imneme! I've updated the PR

**Comment by ph4r05** [comment]
fixed with #68

**Comment by ph4r05** [comment]
@imneme thanks for merging! I am already using the merged master and it works great for me. 

Just for completeness, as I was trying to modify our SeedSeq, I stumbled upon [documentation](https://en.cppreference.com/w/cpp/named_req/SeedSequence) and it says:

> q.generate(rb,re) | void | Fills [rb,re) with 32-bit quantities depending on the initial supplied values and potential previous calls to generate. If rb == re, it does nothing.

Also according to the [generate() documentation](https://en.cppreference.com/w/cpp/numeric/random/seed_seq/generate) the output depends on the size and input seed. Same-sized arrays get the same content. I would say there is a possibility that the algorithm for `generate()` could change in the future and it may lose same-sized-same-content principle property.

I've checked that arrays of different sizes get different seeds. With 
```cpp 
td::seed_seq seq { 42, 54 }; 
```

The sequences of 1, 2, 3 are:
```
4037248171
3007093135, 2886989222
768751226, 3669117059, 3116470141, 627090947
```

The current master is protected from such change - thanks :)
